src/EuroDNS/Helper/EuroDNSApi.php

```diff
@@ -0,0 +1,1793 @@
+     */
+    private function setNs($action)
+    {
+        $i = 1;
+        $ns = null;
+
+        // Check if the action is 'update'
+        if ($action == 'update') {
+            $nameserversArray = $this->nameServers;
+        } else {
+            // Decode the JSON and retrieve name servers from parameters
+            $nameserversArray = json_decode(json_encode($this->params['nameservers']), true);
+        }
+
+        // Loop through name servers and generate XML
+        foreach ($nameserversArray as $key => $val) {
+
+            $host = isset($val['host'])? $val['host']: $val;
+
+            // Check if the host is not empty
+            if (!empty($host)) {
+                // Generate XML for the name server
+                $ns .= "
+                    <nameserver:{$action}>
+                        <nameserver:priority>{$i}</nameserver:priority>
+                        <nameserver:fqdn>{$host}</nameserver:fqdn>
+                        <nameserver:ipaddr>" . gethostbyname($host) . "</nameserver:ipaddr>
+                    </nameserver:{$action}>";
+            }
+
+            $i++;
+        }
+
+        return $ns;
+    }
+
+
+    /**
+     * Set contact details for the specified action.
+     *
+     * @param string $action      The action to perform ('create', 'update', etc.).
+     *
+     * @return string $contact    The formatted XML for contact details.
+     */
+    private function setContact($action)
+    {
+        // Retrieve contact details from billing, registrant, admin, and tech sections
+        $billingDetails = $this->params['billing']['register'];
+        $registrantDetails = $this->params['registrant']['register'];
+        $techDetails = $this->params['tech']['register'];
+        $adminDetails = $this->params['admin']['register'];
+
+        // Replace '+' with '00' in phone numbers to avoid API passing errors
+        $billingPhone = str_replace('+', '00', $billingDetails['phone']);
+        $contactPhone = str_replace('+', '00', $registrantDetails['phone']);
+        $adminPhone = str_replace('+', '00', $adminDetails['phone']);
+        $techPhone = str_replace('+', '00', $techDetails['phone']);
+
+        //find firstname and last name from the given name
+        $nameBilling = $this->splitName($billingDetails['name']);
+        $nameContact = $this->splitName($registrantDetails['name']);
+        $nameAdmin= $this->splitName($adminDetails['name']);
+        $nameTech = $this->splitName($techDetails['name']);
+
+        // Build the XML for contact details
+        $contact = "
+            <contact:{$action}>
+                <contact:type>billing</contact:type>
+                <contact:firstname>{$nameBilling['first_name']}</contact:firstname>
+                <contact:lastname>{$nameBilling['last_name']}</contact:lastname>
+                <contact:company>{$billingDetails['organisation']}</contact:company>
+                <contact:address1>{$billingDetails['address1']}</contact:address1>
```

Comment:
Whilst i like the simplicity of interpolating variables into an inline xml template, it's crucial that these are always escaped so that if the variables contain any special characters, this doesn't cause any problems. Using PHP's built-in `htmlspecialchars()` function should do the trick since XML and HTML work with the same characterspace.

Reply:
updated it like  <contact:firstname>".htmlspecialchars($nameTech['first_name'], ENT_QUOTES, 'UTF-8')."</contact:firstname>